pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
+    >>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
+    >>> jpc = DeviceJacobians(device, {}, config)
+
+    This same class can also be used with the old device interface.
+
+    >>> device = qml.device('lightning.qubit', wires=5)
+    >>> gradient_kwargs = {"method": "adjoint_jacobian"}
+    >>> jpc_lightning = DeviceJacobians(device, gradient_kwargs)
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and jacobian for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents, the
+    caching and retrieval should be more performant than a lookup based on the potentially expensive ``QuantumScript.hash``.
+    This means that the batch of tapes must be the
+    same instance, not just something that looks the same but has a different location in memory.
+
+    When a forward pass with :meth:`~.execute` is called, both the results and the jacobian for the object are stored.
+
+    >>> tape = qml.tape.QuantumScript([qml.RX(1.0, wires=0)], [qml.expval(qml.PauliZ(0))])
+    >>> batch = (tape, )
+    >>> with device.tracker:
+    ...     results = jpc.execute(batch )
+    >>> results
+    (0.5403023058681398,)
+    >>> device.tracker.totals
+    {'execute_and_derivative_batches': 1, 'executions': 1, 'derivatives': 1}
+    >>> jpc._jacs_cache
+    LRUCache({5660934048: (array(-0.84147098),)}, maxsize=10, currsize=1)
+
+    Then when the vjp, jvp, or jacobian is requested, that cached value is used instead of requesting from
+    the device again.
+
+    >>> with device.tracker:
+    ...     vjp = jpc.compute_vjp(batch , (0.5, ) )
+    >>> vjp
+    (array([-0.42073549]),)
+    >>> device.tracker.totals
+    {}
+
+    **Lack of shot vector suppoprt:**
+
+    Given we currently do not have any examples of device derivatives with shots vectors (or even finite shots),
+    we are bypassing support for partitioned shots.
+
+    """
+
+    def __repr__(self):
+        return f"<DeviceJacobians: {self._device.name}, {self._gradient_kwargs}, {self._execution_config}>"
+
+    def __init__(
+        self,
+        device: Union[qml.devices.Device, qml.Device],
+        gradient_kwargs: dict,
+        execution_config: Optional["qml.devices.ExecutionConfig"] = None,
+    ):
+        if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+            logger.debug(
+                "DeviceJacobians created with (%s, %s, %s)",
+                device,
+                execution_config,
+                gradient_kwargs,
+            )
+
+        self._device = device
+        self._execution_config = execution_config
+        self._gradient_kwargs = gradient_kwargs
+
+        self._is_new_device = not isinstance(device, qml.Device)
```

Comment:
Not a major thing but while reviewing the tests I noticed that `jpc._is_new_device` reads a bit funny if `jpc` is not a device. Maybe 
```suggestion
        self._uses_new_device = not isinstance(device, qml.Device)
```